[cmake] Improve generated Visual Studio and Xcode projects

fetzerch opened the pull request:
This PR changes the CMake build system so that usable Visual Studio and Xcode projects are generated.

The initial project files created one static library per subdirectory in xbmc/ (similar to our Autotools project).
While this has some advantages (compile options and include dirs can be set on a directory level, link time is faster), it doesn't play well with the generated Visual Studio and Xcode projects. These treat all subdirectories as separate subprojects and the IDEs project browsers weren't close to the filesystem anymore and rather unusable. Additionally the subprojects caused much longer compile times especially with Visual Studio.

Screenshots can be found in Jira: http://jira.kodi.tv/browse/KODI-16651

Pings for code review: @wsnipex, @notspiff, @hudokkow 
Would be good to get some feedback from Xcode users: @FernetMenta, @Memphiz

Jenkins build: http://jenkins.kodi.tv/view/Helpers/job/BuildMulti-All/1503/
Packages should have been uploaded in case platform devs want to verify that they are still working.

fetzerch committed:
[cmake] Move enabling of assembler to main CMakeLists.txt

When we use the option to not build static libs per subdir, we cannot
enable assembler only in a subfolder.

Change is non intrusive it just verifies compiler support for asm
earlier in the chain.

fetzerch committed:
[cmake/win32] Use C++ compiler to compile .c files

Move setting the LANGUAGE source file properties into a function.
This is in preparation to optionally build static libraries.
This property can only be specified in the CMakeLists.txt that
generates the target. Later in the non static lib case, it will be
called in the main executable/library.

fetzerch committed:
[cmake/win32] Disable TestCryptThreading

The CryptThreading class is not meant to be used on Windows.

fetzerch committed:
[cmake] Group sources and targets into folders

fetzerch committed:
[cmake] Introduce ENABLE_STATIC_LIBS option

Add an option ENABLE_STATIC_LIBS that defines if the buildsystem
creates one static library per subdirectory in xbmc/ or compiles
everything into one big library.

The option defaults to true, except for Visual Studio and Xcode,
because these generators work much better with a single big library.
The project structure in Visual Studio and Xcode is closer to the
filesystem, which cannot be realized otherwise, because they show
separate projects for each subdirectory.

Several related changes are introduced in this commit:
- dependency management had to be adapted to work in both cases
- include directories are propagated through
  target_include_directories(... INTERFACE ...)
- dvdnav and crossguid are made a project wide dependency, because they
  are widely used within the project.

FernetMenta commented:
@fetzerch any significant changes to the version I tested last week? Are you asking for general feedback or do you want to have this particular version tested?

fetzerch commented:
@FernetMenta you mentioned some build errors when only building the kodi scheme. That should work now. Other than that, just some minor improvements.

In general I'm interested in anything that prevents you from using it

hudokkow commented on project/cmake/scripts/common/Macros.cmake:
```diff
@@ -176,6 +177,20 @@ function(core_add_shared_library name)
 
+# Sets the compile language for all C source files in a target to CXX.
+# Needs to be called from the CMakeLists.txt that defines the target.
+# This method needs to be called from the CMakeLists.txt that adds the target!
```
A bit redundant? Or shall we all be scared? :smile:

fetzerch replied:
Hehe, good catch :)

hudokkow commented:
Quick browse and code looks good but I can't test Xcode stuff.
I can test VS on Monday if I get past the "install'uninstall one more thing because we're so great and it will only take 3h to download, verify, spit an error and tell you this is the wrong install/update exec. Or we won't tell you. We're M$ and we know you have psychic powers!"

wsnipex commented on project/cmake/CMakeLists.txt:
```diff
@@ -293,24 +301,43 @@ elseif(CORE_SYSTEM_NAME STREQUAL android)
 endif()
 whole_archive(_MAIN_LIBRARIES ${core_DEPENDS})
-target_link_libraries(${APP_NAME_LC} ${SYSTEM_LDFLAGS} ${_MAIN_LIBRARIES} ${DEPLIBS} ${CMAKE_DL_LIBS})
+target_link_libraries(${APP_NAME_LC} lib${APP_NAME_LC} ${_MAIN_LIBRARIES} lib${APP_NAME_LC} ${SYSTEM_LDFLAGS} ${DEPLIBS} ${CMAKE_DL_LIBS})
```
double lib${APP_NAME_LC}

fetzerch replied:
thx, fixed

fetzerch commented:
Currently Kodi won't launch if built on windows with CMake and Visual Studio because there is a mismatch between where the dlls are downloaded/built to and where Kodi tries to load them. This got introduced by: https://github.com/xbmc/xbmc/commit/e3391adb4f74b0a13de1e76a967fd604771ad6e8

fetzerch committed:
[cmake] Add a 'destination directory' parameter to copy_file_to_buildtree

The function copy_file_to_buildtree is used to mirror files to the
build directory, so that Kodi can be launched from there.

The previous implementation of that function only allowed to mirror
files to the same tree structure. With this change files can be
mirrored also into different directories as this is needed for some
use cases (DLLs on Windows, certificates on IOS/OSX).

This changes the following parts:
- Patterns specified in installdata now support an optional
  second parameter that specifies where the files should be mirrored to.
  This parameter is optional. If specified it points to the destination
  directory (relative to CMAKE_BINARY_DIR) otherwise the files are
  mirrored in the same tree structure under CMAKE_BINARY_DIR.
- Mirror files for in-source-builds where necessary.
- Remove the relative parameter from copy_file_to_buildtree.
  Its API was confusing and it's not necessary because the files are
  always relative to CORE_SOURCE_DIR.
- Remove Windows specific hook to mirror additional files and
  use the newly introduced mechanism instead.

fetzerch committed:
[cmake] Install openssl certs.pem for OSX and IOS

fetzerch committed:
[cmake/win32] Use precompiled cpluff

fetzerch committed:
[cmake/win32] Shared libraries need to be generated into the build root folder

While shared libraries have to be built into the system folder on most
platforms, on Windows they have to go directly into the root build folder.

fetzerch commented:
Pushed more commits:
- ccc1ded, 9778be8 and ccebec9 fix the problem of kodi not starting on windows
- ccc1ded together with 40702e8 replace https://github.com/xbmc/xbmc/pull/10162.
  @Memphiz: Do you want me to exclude the commit installing certs.pem on osx/ios here? From what I can see it doesn't cause any harm to have it even if #10162 is not yet included.

Memphiz commented:
no need to exclude it - once i have finally some time to work on that PR i will do. The packaging of the cert will remain as is - so would be good to have it already in cmake ...

fetzerch commented:
Build was fine except on linux (caused by PulseAudio incompatibility).
New build triggered after fixing that is green: http://jenkins.kodi.tv/job/LINUX-64/8479/console

The feedback I got so far was positive, so I'll merge it now. Further improvements can be done independently.

FernetMenta commented:
> The feedback I got so far was positive, so I'll merge it now

only fixes can be merged outside a merge window. please keep in mind next time

fetzerch commented:
Perfectly clear to me. However windows build was broken without this due to libraries being moved around. Additionally I understood from @MartijnKaijser that cmake things can go in since they make our live easier and have little to no impact on behavior. Otherwise I wouldn't have merged it.

MartijnKaijser commented:
This code is currently unused by our buildsystem/official builds as such it isn't really affected by merge windows. This will of course change when we start using it.

FernetMenta commented:
ok, thanks for the update
